src/Shell32.Desktop/Shell32.cs

```diff
@@ -0,0 +1,111 @@
+        /// </remarks>
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for SHGetKnownFolderPath.")]
+        [DllImport(nameof(Shell32), SetLastError = true)]
+        public static extern unsafe HResult SHGetFolderPath(
+            IntPtr hwndOwner,
+            CSIDL nFolder,
+            IntPtr hToken,
+            SHGetFolderPathFlags dwFlags,
+            out IntPtr pszPath);
+
+        /// <summary>
+        /// Gets the path of a folder identified by a CSIDL value.
+        /// </summary>
+        /// <param name="rfid">
+        /// A <see cref="Guid"/> value that identifies the folder whose path is to be retrieved.
+        /// </param>
+        /// <param name="dwFlags">Flags that specify special retrieval options. This value can be 0; otherwise, one or more of the <see cref="KNOWN_FOLDER_FLAG"/> values.</param>
+        /// <param name="hToken">
+        /// An access token that represents a particular user. If this parameter is NULL, which is the most common usage, the function requests the known folder for the current user.
+        /// Assigning the hToken parameter a value of -1 indicates the Default User.
+        /// 
+        /// Microsoft Windows 2000 and earlier: Always set this parameter to NULL.
+        /// Windows XP and later: This parameter is usually set to NULL, but you might need to assign a non-NULL value to hToken for those folders that can have multiple users but are treated as belonging to a single user.The most commonly used folder of this type is Documents.
+        /// </param>
+        /// <param name="pszPath">
+        /// A pointer to a null-terminated string of length MAX_PATH which will receive the path.
+        /// If an error occurs or <see cref="HResult.Code.S_FALSE"/> is returned, this string will be empty.
+        /// </param>
+        /// <returns>If this function succeeds, it returns <see cref="HResult.Code.S_OK"/>. Otherwise, it returns an <see cref="HResult"/> error code.</returns>
+        /// <remarks>
+        /// As of Windows Vista, this function is merely a wrapper for <see cref="SHGetKnownFolderPath"/>.
+        /// The returned path does not include a trailing backslash.
+        /// For example, "C:\Users" is returned rather than "C:\Users\".
+        /// </remarks>
+        [DllImport(nameof(Shell32), SetLastError = true)]
+        public static extern unsafe HResult SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            KNOWN_FOLDER_FLAG dwFlags,
+            IntPtr hToken,
+            out IntPtr pszPath);
+
+        [DllImport(nameof(Shell32), SetLastError = true)]
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for SHGetKnownFolderIDList")]
+        public static extern unsafe HResult SHGetFolderLocation(
+            IntPtr hwndOwner,
+            CSIDL nFolder,
+            IntPtr hToken,
+            int dwReserved,
+            out IntPtr pidl);
```

Comment:
I don't think i'll have time to do it before this weekend but with something approaching what I posted in my comment to `SHITEMID`, `ITEMIDLIST` itself shouldn't be problematic.

Regarding pointers, I don't know how @AArnott see it but I see this library as an attempt to get the correct types as much as we can, `SafeHandle` for opaque types having a custom destruction function, pointers instead of `IntPtr`... 
That's more complex than what is present on pinvoke.net, and there is often a big step of refactorings when importing pinvokes from an existing code base. The target isn't to have a NuGet version of pinvoke.net (I still wonder why they never did it, it would be relatively simple to create) but to work on getting a clean API surface that resemble as much as possible the C version, adding helpers to ease C# consumption.